components/fxa-client/src/internal/state_manager.rs

```diff
@@ -157,23 +169,64 @@ impl StateManager {
     ) {
         // When our keys change, we might need to re-register device capabilities with the server.
         // Ensure that this happens on the next call to ensure_capabilities.
-        self.persisted_state.device_capabilities.clear();
+        self.clear_server_local_device_info();
 
         for (scope, key) in scoped_keys {
             self.persisted_state.scoped_keys.insert(scope, key);
         }
         self.persisted_state.refresh_token = Some(refresh_token);
         self.persisted_state.session_token = session_token;
+        self.persisted_state.logged_out_from_auth_issues = false;
         self.flow_store.clear();
     }
 
     /// Called when the account is disconnected.  This clears most of the auth state, but keeps
     /// some information in order to eventually reconnect to the same user account later.
     pub fn disconnect(&mut self) {
-        self.persisted_state = self.persisted_state.start_over();
+        // Clear (almost all of) the persisted state of the account.
+        //
+        // This method keep just enough information to be able to eventually reconnect
+        // to the same user account later. To completely forget the previously-signed-in
+        // user, simply discard the persisted data.
+
+        self.persisted_state.current_device_id = None;
+        self.persisted_state.refresh_token = None;
+        self.persisted_state.scoped_keys = HashMap::new();
+        self.persisted_state.last_handled_command = None;
+        self.persisted_state.commands_data = HashMap::new();
+        self.persisted_state.access_token_cache = HashMap::new();
+        self.persisted_state.device_capabilities = HashSet::new();
+        self.persisted_state.server_local_device_info = None;
+        self.persisted_state.session_token = None;
+        self.persisted_state.logged_out_from_auth_issues = false;
+        self.flow_store.clear();
+    }
+
+    /// Called when the account is logged out because of auth issues.  This clears some of the auth
+    /// state, but less than disconnect().
```

Comment:
what do you think about being explicit what data is not deleted (device id, right?)